--- obexd-0.40/client/sync.c	2011-04-25 13:28:37.436981087 +0800
+++ new/obexd-0.40/client/sync.c	2011-04-25 13:28:19.339950617 +0800
@@ -78,6 +78,9 @@
 	DBusMessage *reply;
 	char *buf = NULL;
 
+	if (session->msg == NULL)
+		goto done;
+
 	reply = dbus_message_new_method_return(session->msg);
 
 	if (transfer->filled > 0)
@@ -91,6 +94,8 @@
 	g_dbus_send_message(session->conn, reply);
 	dbus_message_unref(session->msg);
 	session->msg = NULL;
+done:
+	transfer_unregister(transfer); 
 }
 
 static DBusMessage *sync_getphonebook(DBusConnection *connection,

--- old/plugins/irmc.c	2011-07-15 15:32:17.448664952 +0800
+++ new/plugins/irmc.c	2011-07-15 15:37:20.548581935 +0800
@@ -167,6 +167,8 @@
 	else
 		irmc->buffer = g_string_append(irmc->buffer, owner_vcard);
 
+	if (buffer == NULL)
+		goto finish;
 	/* loop around buffer and add X-IRMC-LUID attribs */
 	s = buffer;
 	while ((t = strstr(s, "UID:")) != NULL) {
@@ -189,7 +191,7 @@
 	}
 	/* add remaining bit of buffer */
 	irmc->buffer = g_string_append(irmc->buffer, s);
-
+finish:
 	obex_object_set_io_flags(irmc, G_IO_IN, 0);
 }

