diff --git a/gobex/gobex.c b/gobex/gobex.c
index 7c136af..104c22e 100644
--- a/gobex/gobex.c
+++ b/gobex/gobex.c
@@ -1295,7 +1295,7 @@ GObex *g_obex_new(GIOChannel *io, GObexTransportType transport_type,
 		obex->write = write_stream;
 		break;
 	case G_OBEX_TRANSPORT_PACKET:
-		obex->use_srm = TRUE;
+		/* obex->use_srm = TRUE; */
 		obex->read = read_packet;
 		obex->write = write_packet;
 		break;
