diff -Naur obexd.orig/plugins/pbap.c obexd/plugins/pbap.c
--- obexd.orig/plugins/pbap.c	2014-11-05 10:15:09.025110346 +0200
+++ obexd/plugins/pbap.c	2014-11-05 10:15:22.149110833 +0200
@@ -714,8 +714,13 @@
 
 	manager_unregister_session(os);
 
-	if (pbap->obj)
+	if (pbap->obj) {
+		if (pbap->obj->request) {
+			phonebook_req_finalize(pbap->obj->request);
+			pbap->obj->request = NULL;
+		}
 		pbap->obj->session = NULL;
+	}
 
 	if (pbap->params) {
 		g_free(pbap->params->searchval);
@@ -818,14 +823,20 @@
 	if (obj->session)
 		obj->session->obj = NULL;
 
-	if (obj->buffer)
+	if (obj->buffer) {
 		g_string_free(obj->buffer, TRUE);
+		obj->buffer = NULL;
+	}
 
-	if (obj->apparam)
+	if (obj->apparam) {
 		g_obex_apparam_free(obj->apparam);
+		obj->apparam = NULL;
+	}
 
-	if (obj->request)
+	if (obj->request) {
 		phonebook_req_finalize(obj->request);
+		obj->request = NULL;
+	}
 
 	g_free(obj);
 
diff -Naur obexd.orig/plugins/phonebook-sailfish.c obexd/plugins/phonebook-sailfish.c
--- obexd.orig/plugins/phonebook-sailfish.c	2014-11-05 10:15:09.025110346 +0200
+++ obexd/plugins/phonebook-sailfish.c	2014-11-05 10:55:32.717200369 +0200
@@ -676,6 +676,8 @@
 				cb,
 				data,
 				NULL);
+	dbus_message_unref(msg);
+
 	return 0;
 }
 
@@ -749,6 +751,7 @@
 				fetch_one_cb, 
 				data,
 				NULL);
+	dbus_message_unref(msg);
 
 	if (err)
 		*err = 0;
@@ -829,6 +832,7 @@
 				fetch_many_cb,
 				data,
 				NULL);
+	dbus_message_unref(msg);
 
 	if (err)
 		*err = 0;
@@ -845,8 +849,10 @@
 	if (!data)
 		return;
 
-	if (data->pend != NULL)
+	if (data->pend != NULL) {
 		dbus_pending_call_cancel(data->pend);
+		dbus_pending_call_unref(data->pend);
+	}
 
 	g_free(data->pull_buf);
 	g_free(data->name);
